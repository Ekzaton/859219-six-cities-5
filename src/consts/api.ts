export const BASE_URL = `https://6.react.pages.academy/six-cities`;
export const REQUEST_TIMEOUT = 5000;
