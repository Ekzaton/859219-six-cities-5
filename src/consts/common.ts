export enum AppRoute {
  MAIN = `/`,
  LOGIN = `/login/`,
  FAVORITES = `/favorites/`,
  OFFER_ID = `/offer/:id/`,
  OFFER = `/offer/`
}

export enum AuthStatus {
  AUTH = `auth`,
  NO_AUTH = `no-auth`
}

export enum FilteringType {
  PARIS = `Paris`,
  COLOGNE = `Cologne`,
  BRUSSELS = `Brussels`,
  AMSTERDAM = `Amsterdam`,
  HAMBURG = `Hamburg`,
  DUSSELDORF = `Dusseldorf`
}

export enum SortingType {
  POPULAR = `Popular`,
  PRICE_LOW_TO_HIGH = `Price: low to high`,
  PRICE_HIGH_TO_LOW = `Price: high to low`,
  TOP_RATED_FIRST = `Top rated first`
}
