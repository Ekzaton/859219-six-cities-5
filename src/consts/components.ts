export const OFFER_IMAGES_COUNT = 6;
export const OFFER_REVIEWS_COUNT = 10;
export const RATING_STARS_COUNT = 5;
export const REG_EXP_EMAIL = /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/;
export const REG_EXP_PASSWORD = /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/;
export const DELAY_DURATION = 250;

export enum BtnBigSize {
  WIDTH = 31,
  HEIGHT = 33
}

export enum BtnSize {
  WIDTH = 18,
  HEIGHT = 19
}

export enum BtnType {
  PLACE_CARD = `place-card`,
  PROPERTY = `property`
}

export enum CardImgSize {
  WIDTH = 260,
  HEIGHT = 200
}

export enum CardType {
  CITIES = `cities`,
  FAVORITES = `favorites`,
  NEAR_PLACES = `near-places`
}

export enum FavCardImgSize {
  WIDTH = 150,
  HEIGHT = 110
}

export enum HeaderType {
  MAIN = `main`,
  LOGIN = `login`
}

export enum MapIconSize {
  WIDTH = 30,
  HEIGHT = WIDTH
}

export enum MapIconUrl {
  PIN = `img/pin.svg`,
  PIN_ACTIVE = `img/pin-active.svg`
}

export enum MapLayer {
  TEMPLATE = `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
  ATTRIBUTION = `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
}

export enum MapType {
  CITIES = `cities`,
  PROPERTY = `property`
}

export enum ReviewDate {
  LOCALE = `en-US`,
  MONTH = `long`,
  YEAR = `numeric`
}

export enum ReviewLehgth {
  MIN = 50,
  MAX = 300
}
