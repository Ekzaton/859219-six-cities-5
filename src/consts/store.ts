export enum ActionType {
  CHANGE_ACTIVE_OFFER_ID = `CHANGE_ACTIVE_OFFER_ID`,
  CHANGE_CURRENT_CITY = `CHANGE_CURRENT_CITY`,
  CHANGE_CURRENT_SORTING = `CHANGE_CURRENT_SORTING`,
  CHANGE_OFFER_FAVORITE_STATUS = `CHANGE_OFFER_FAVORITE_STATUS`,
  GET_ALL_OFFERS = `GET_ALL_OFFERS`,
  GET_FAVORITE_OFFERS = `GET_FAVORITE_OFFERS`,
  GET_NEARBY_OFFERS = `GET_NEARBY_OFFERS`,
  GET_SINGLE_OFFER = `GET_SINGLE_OFFER`,
  GET_REVIEWS = `GET_REVIEWS`,
  GET_USER = `GET_USER`,
  REDIRECT_TO_ROUTE = `REDIRECT_TO_ROUTE`,
  SET_AUTH_STATUS = `SET_AUTH_STATUS`,
  SET_INITIAL_STATE = `SET_INITIAL_STATE`,
  SET_IS_ALL_OFFERS_LOADING = `SET_IS_ALL_OFFERS_LOADING`,
  SET_IS_ALL_OFFERS_LOADING_ERROR = `SET_IS_ALL_OFFERS_LOADING_ERROR`,
  SET_IS_FAVORITE_OFFERS_LOADING = `SET_IS_FAVORITE_OFFERS_LOADING`,
  SET_IS_FAVORITE_OFFERS_LOADING_ERROR = `SET_IS_FAVORITE_OFFERS_LOADING_ERROR`,
  SET_IS_FAVORITE_SENDING = `SET_IS_FAVORITE_SENDING`,
  SET_IS_FAVORITE_SENDING_ERROR = `SET_IS_FAVORITE_SENDING_ERROR`,
  SET_IS_LOGIN_SENDING = `SET_IS_LOGIN_SENDING`,
  SET_IS_LOGIN_SENDING_ERROR = `SET_IS_LOGIN_SENDING_ERROR`,
  SET_IS_REVIEW_SENDING = `SET_IS_REVIEW_SENDING`,
  SET_IS_REVIEW_SENDING_ERROR = `SET_IS_REVIEW_SENDING_ERROR`,
  SET_IS_SINGLE_OFFER_LOADING = `SET_IS_SINGLE_OFFER_LOADING`,
  SET_IS_SINGLE_OFFER_LOADING_ERROR = `SET_IS_SINGLE_OFFER_LOADING_ERROR`
}

export enum APIEndpoint {
  HOTELS = `/hotels/`,
  NEARBY = `/nearby/`,
  FAVORITE = `/favorite/`,
  COMMENTS = `/comments/`,
  LOGIN = `/login/`,
  LOGOUT = `/logout/`
}
