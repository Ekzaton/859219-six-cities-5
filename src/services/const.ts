export const BASE_URL = `https://5.react.pages.academy/six-cities`;
export const REQUEST_TIMEOUT = 5000;

export enum HTTPCode {
  BAD_REQUEST = 400,
  UNAUTHORIZED = 401,
  NOT_FOUND = 404
}
