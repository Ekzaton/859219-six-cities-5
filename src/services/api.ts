import axios, {AxiosError, AxiosInstance, AxiosResponse} from "axios";
import camelcaseKeys from "camelcase-keys";

import {BASE_URL, REQUEST_TIMEOUT} from "../consts/api";

export const createAPI = (): AxiosInstance => {
  const api = axios.create({
    baseURL: BASE_URL,
    timeout: REQUEST_TIMEOUT,
    withCredentials: true,
  });

  const onSuccess = (response: AxiosResponse) => {
    return {...response, data: camelcaseKeys(response.data, {deep: true})};
  };

  const onFail = (error: AxiosError) => {
    throw error;
  };

  api.interceptors.response.use(onSuccess, onFail);

  return api;
};
