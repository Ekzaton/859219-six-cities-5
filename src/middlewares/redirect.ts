import {Middleware} from "redux";

import {ActionType} from "../consts/store";

import history from "../history/history";

export const redirect: Middleware = () => (dispatch) => (action) => {
  if (action.type === ActionType.SET_REDIRECT_TO_ROUTE) {
    history.push(action.payload);
  }

  return dispatch(action);
};
