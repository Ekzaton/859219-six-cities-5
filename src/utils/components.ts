import {ReviewDate, ReviewLehgth} from "../const";

export const getOfferType = (type: string): string => type[0].toUpperCase() + type.slice(1);

export const getRatingArray = (length: number): number[] => new Array(length).fill(``);

export const getRatingStars = (rating: number): string => `${rating * 20}%`;

export const getReviewDate = (date: string): string =>
  new Date(date).toLocaleString(ReviewDate.LOCALE, {month: ReviewDate.MONTH, year: ReviewDate.YEAR});

export const validateComment = (comment: string): boolean =>
  comment.length >= ReviewLehgth.MIN && comment.length <= ReviewLehgth.MAX;


