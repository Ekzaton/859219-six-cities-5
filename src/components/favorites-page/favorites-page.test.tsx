import React from 'react';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {act, render} from '@testing-library/react';

import {AuthStatus} from "../../consts/common";
import {DELAY_DURATION} from "../../consts/components";
import {redirect} from "../../middlewares/redirect";
import offersMock from "../../mock/offers.json";
import userMock from "../../mock/user.json";
import {createAPI} from "../../services/api";

import FavoritesPage from "./favorites-page";

describe(`Favorites page renders correctly`, () => {
  const api = createAPI();
  const thunkWithAPI = thunk.withExtraArgument(api);
  const mockStore = createMockStore([thunkWithAPI, redirect]);

  it(`for authorized user`, () => {
    const store = mockStore({
      favorites: {
        favoriteOffers: offersMock,
        isLoading: false
      },
      login: {
        authStatus: AuthStatus.AUTH,
        user: userMock
      }
    });

    const {container} = render(
        <Provider store={store}>
          <MemoryRouter>
            <FavoritesPage/>
          </MemoryRouter>
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it(`for authorized user (no offers)`, () => {
    const store = mockStore({
      favorites: {
        favoriteOffers: [],
        isLoading: false
      },
      login: {
        authStatus: AuthStatus.AUTH,
        user: userMock
      }
    });

    jest.useFakeTimers();

    const {container} = render(
        <Provider store={store}>
          <MemoryRouter>
            <FavoritesPage/>
          </MemoryRouter>
        </Provider>
    );

    act(() => {
      jest.advanceTimersByTime(DELAY_DURATION);
    });

    expect(container).toMatchSnapshot();
  });

  it(`for authorized user (loading)`, () => {
    const store = mockStore({
      favorites: {
        favoriteOffers: [],
        isLoading: true
      },
      login: {
        authStatus: AuthStatus.AUTH,
        user: userMock
      }
    });

    jest.useFakeTimers();

    const {container} = render(
        <Provider store={store}>
          <MemoryRouter>
            <FavoritesPage/>
          </MemoryRouter>
        </Provider>
    );

    act(() => {
      jest.advanceTimersByTime(DELAY_DURATION);
    });

    expect(container).toMatchSnapshot();
  });
});
