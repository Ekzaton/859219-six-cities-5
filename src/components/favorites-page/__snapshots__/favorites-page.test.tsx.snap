// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites page renders correctly for authorized user (loading) 1`] = `
<div>
  <main
    class="page__main"
  >
    <h1
      class="property__name"
      style="margin-top: 45vh;"
    >
      Loading data... Please wait
    </h1>
  </main>
</div>
`;

exports[`Favorites page renders correctly for authorized user (no offers) 1`] = `
<div>
  <div
    class="page"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <div
          class="header__wrapper"
        >
          <div
            class="header__left"
          >
            <a
              class="header__logo-link"
              href="/"
            >
              <img
                alt="6 cities logo"
                class="header__logo"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            class="header__nav"
          >
            <ul
              class="header__nav-list"
            >
              <li
                class="header__nav-item user"
              >
                <a
                  class="header__nav-link header__nav-link--profile"
                  href="/favorites/"
                >
                  <div
                    class="header__avatar-wrapper user__avatar-wrapper"
                    style="background-image: url(img/avatar-max.jpg); border-radius: 50%;"
                  />
                  <span
                    class="header__user-name user__name"
                  >
                    user-1@mail.ru
                  </span>
                </a>
              </li>
              <li
                class="header__nav-item"
              >
                <a
                  class="header__nav-link"
                  style="margin-left: 10px;"
                >
                  <span
                    class="header__login"
                  >
                    [Sign out]
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      class="page__main page__main--favorites page__main--favorites-empty"
    >
      <div
        class="page__favorites-container container"
      >
        <section
          class="favorites favorites--empty"
        >
          <h1
            class="visually-hidden"
          >
            Favorites (empty)
          </h1>
          <div
            class="favorites__status-wrapper"
          >
            <b
              class="favorites__status"
            >
              Nothing yet saved.
            </b>
            <p
              class="favorites__status-description"
            >
              Save properties to narrow down search or plan yor future trips.
            </p>
          </div>
        </section>
      </div>
    </main>
    <footer
      class="footer container"
    >
      <a
        class="footer__logo-link"
        href="/"
      >
        <img
          alt="6 cities logo"
          class="footer__logo"
          height="33"
          src="img/logo.svg"
          width="64"
        />
      </a>
    </footer>
  </div>
</div>
`;

exports[`Favorites page renders correctly for authorized user 1`] = `
<div>
  <div
    class="page"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <div
          class="header__wrapper"
        >
          <div
            class="header__left"
          >
            <a
              class="header__logo-link"
              href="/"
            >
              <img
                alt="6 cities logo"
                class="header__logo"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            class="header__nav"
          >
            <ul
              class="header__nav-list"
            >
              <li
                class="header__nav-item user"
              >
                <a
                  class="header__nav-link header__nav-link--profile"
                  href="/favorites/"
                >
                  <div
                    class="header__avatar-wrapper user__avatar-wrapper"
                    style="background-image: url(img/avatar-max.jpg); border-radius: 50%;"
                  />
                  <span
                    class="header__user-name user__name"
                  >
                    user-1@mail.ru
                  </span>
                </a>
              </li>
              <li
                class="header__nav-item"
              >
                <a
                  class="header__nav-link"
                  style="margin-left: 10px;"
                >
                  <span
                    class="header__login"
                  >
                    [Sign out]
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      class="page__main page__main--favorites false"
    >
      <div
        class="page__favorites-container container"
      >
        <section
          class="favorites"
        >
          <h1
            class="favorites__title"
          >
            Saved listing
          </h1>
          <ul
            class="favorites__list"
          >
            <li
              class="favorites__locations-items"
            >
              <div
                class="favorites__locations locations locations--current"
              >
                <div
                  class="locations__item"
                >
                  <a
                    class="locations__item-link"
                    href="/"
                  >
                    <span>
                      Amsterdam
                    </span>
                  </a>
                </div>
              </div>
              <div
                class="favorites__places"
              >
                <article
                  class="favorites__card place-card"
                >
                  <div
                    class="favorites__image-wrapper place-card__image-wrapper"
                  >
                    <a
                      href="/offer/1"
                    >
                      <img
                        alt="Place image"
                        class="place-card__image"
                        height="110"
                        src="img/apartment-01.jpg"
                        width="150"
                      />
                    </a>
                  </div>
                  <div
                    class="favorites_card-info place-card__info"
                  >
                    <div
                      class="place-card__price-wrapper"
                    >
                      <div
                        class="place-card__price"
                      >
                        <b
                          class="place-card__price-value"
                        >
                          €
                          120
                        </b>
                        <span
                          class="place-card__price-text"
                        >
                           / night
                        </span>
                      </div>
                      <button
                        class="place-card__bookmark-button false button"
                        type="button"
                      >
                        <svg
                          class="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlink:href="#icon-bookmark"
                          />
                        </svg>
                        <span
                          class="visually-hidden"
                        >
                          To bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      class="place-card__rating rating"
                    >
                      <div
                        class="place-card__stars rating__stars"
                      >
                        <span
                          style="width: 96%;"
                        />
                        <span
                          class="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      class="place-card__name"
                    >
                      <a
                        href="/offer/1"
                      >
                        Beautiful & luxurious studio at great location
                      </a>
                    </h2>
                    <p
                      class="place-card__type"
                    >
                      Apartment
                    </p>
                  </div>
                </article>
                <article
                  class="favorites__card place-card"
                >
                  <div
                    class="place-card__mark"
                  >
                    <span>
                      Premium
                    </span>
                  </div>
                  <div
                    class="favorites__image-wrapper place-card__image-wrapper"
                  >
                    <a
                      href="/offer/2"
                    >
                      <img
                        alt="Place image"
                        class="place-card__image"
                        height="110"
                        src="img/apartment-02.jpg"
                        width="150"
                      />
                    </a>
                  </div>
                  <div
                    class="favorites_card-info place-card__info"
                  >
                    <div
                      class="place-card__price-wrapper"
                    >
                      <div
                        class="place-card__price"
                      >
                        <b
                          class="place-card__price-value"
                        >
                          €
                          180
                        </b>
                        <span
                          class="place-card__price-text"
                        >
                           / night
                        </span>
                      </div>
                      <button
                        class="place-card__bookmark-button false button"
                        type="button"
                      >
                        <svg
                          class="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlink:href="#icon-bookmark"
                          />
                        </svg>
                        <span
                          class="visually-hidden"
                        >
                          To bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      class="place-card__rating rating"
                    >
                      <div
                        class="place-card__stars rating__stars"
                      >
                        <span
                          style="width: 100%;"
                        />
                        <span
                          class="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      class="place-card__name"
                    >
                      <a
                        href="/offer/2"
                      >
                        Wood and stone place
                      </a>
                    </h2>
                    <p
                      class="place-card__type"
                    >
                      Hotel
                    </p>
                  </div>
                </article>
                <article
                  class="favorites__card place-card"
                >
                  <div
                    class="favorites__image-wrapper place-card__image-wrapper"
                  >
                    <a
                      href="/offer/3"
                    >
                      <img
                        alt="Place image"
                        class="place-card__image"
                        height="110"
                        src="img/apartment-03.jpg"
                        width="150"
                      />
                    </a>
                  </div>
                  <div
                    class="favorites_card-info place-card__info"
                  >
                    <div
                      class="place-card__price-wrapper"
                    >
                      <div
                        class="place-card__price"
                      >
                        <b
                          class="place-card__price-value"
                        >
                          €
                          100
                        </b>
                        <span
                          class="place-card__price-text"
                        >
                           / night
                        </span>
                      </div>
                      <button
                        class="place-card__bookmark-button false button"
                        type="button"
                      >
                        <svg
                          class="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlink:href="#icon-bookmark"
                          />
                        </svg>
                        <span
                          class="visually-hidden"
                        >
                          To bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      class="place-card__rating rating"
                    >
                      <div
                        class="place-card__stars rating__stars"
                      >
                        <span
                          style="width: 80%;"
                        />
                        <span
                          class="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      class="place-card__name"
                    >
                      <a
                        href="/offer/3"
                      >
                        Canal View Prinsengracht
                      </a>
                    </h2>
                    <p
                      class="place-card__type"
                    >
                      House
                    </p>
                  </div>
                </article>
                <article
                  class="favorites__card place-card"
                >
                  <div
                    class="favorites__image-wrapper place-card__image-wrapper"
                  >
                    <a
                      href="/offer/4"
                    >
                      <img
                        alt="Place image"
                        class="place-card__image"
                        height="110"
                        src="img/room.jpg"
                        width="150"
                      />
                    </a>
                  </div>
                  <div
                    class="favorites_card-info place-card__info"
                  >
                    <div
                      class="place-card__price-wrapper"
                    >
                      <div
                        class="place-card__price"
                      >
                        <b
                          class="place-card__price-value"
                        >
                          €
                          70
                        </b>
                        <span
                          class="place-card__price-text"
                        >
                           / night
                        </span>
                      </div>
                      <button
                        class="place-card__bookmark-button place-card__bookmark-button--active button"
                        type="button"
                      >
                        <svg
                          class="place-card__bookmark-icon"
                          height="19"
                          width="18"
                        >
                          <use
                            xlink:href="#icon-bookmark"
                          />
                        </svg>
                        <span
                          class="visually-hidden"
                        >
                          In bookmarks
                        </span>
                      </button>
                    </div>
                    <div
                      class="place-card__rating rating"
                    >
                      <div
                        class="place-card__stars rating__stars"
                      >
                        <span
                          style="width: 84%;"
                        />
                        <span
                          class="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <h2
                      class="place-card__name"
                    >
                      <a
                        href="/offer/4"
                      >
                        Nice, cozy, warm big bed apartment
                      </a>
                    </h2>
                    <p
                      class="place-card__type"
                    >
                      Room
                    </p>
                  </div>
                </article>
              </div>
            </li>
          </ul>
        </section>
      </div>
    </main>
    <footer
      class="footer container"
    >
      <a
        class="footer__logo-link"
        href="/"
      >
        <img
          alt="6 cities logo"
          class="footer__logo"
          height="33"
          src="img/logo.svg"
          width="64"
        />
      </a>
    </footer>
  </div>
</div>
`;
